aten/src/ATen/native/mps/operations/Unique.mm

```diff
@@ -0,0 +1,345 @@
+}
+
+} // namespace mps
+
+std::tuple<Tensor, Tensor, Tensor>
+_unique_impl_mps(const Tensor& self, const bool return_inverse, const bool return_counts, const bool consecutive, c10::optional<int64_t> dimOpt) {
+
+  const Tensor& input = self.contiguous();
+
+  // get flat output size
+  int64_t totalElems = c10::multiply_integers(input.sizes());
+
+  IntArrayRef outputShape = IntArrayRef(totalElems);
+  IntArrayRef inverseIndicesShape = input.sizes();
+  IntArrayRef countsShape = IntArrayRef(totalElems);
+  int64_t dim = dimOpt.has_value() ? maybe_wrap_dim(dimOpt.value(), self.dim()) : 0;
+
+  if (dimOpt.has_value()) {
+    outputShape = input.sizes();
+    inverseIndicesShape = IntArrayRef(input.sizes()[dim]);
+    countsShape = IntArrayRef(input.sizes()[dim]);
+  }
+  if (!return_inverse)
+    inverseIndicesShape = {};
+  if (!return_counts)
+    countsShape = {};
+
+  Tensor output = at::native::empty_mps(outputShape, input.scalar_type(), c10::nullopt, kMPS);
+  Tensor inverse_indices = at::native::empty_mps(inverseIndicesShape, ScalarType::Long, c10::nullopt, kMPS);
+  Tensor counts = at::native::empty_mps(countsShape, ScalarType::Long, c10::nullopt, kMPS);
+  Tensor length = at::native::empty_mps({1}, ScalarType::Int, c10::nullopt, kMPS);
+
+  if (input.numel() == 0) {
+    return std::make_tuple(output, inverse_indices, counts);
+  }
+
+  mps::UniqueCachedGraph *uniqueGraph = mps::getUniqueGraph(input, return_inverse, return_counts, consecutive, dimOpt);
+  mps::runUniqueGraph(uniqueGraph, input, output, inverse_indices, counts, length, return_inverse, return_counts);
+
+  int64_t lengthScalar = length.item<int64_t>() + 1; // length actually holds max index, add 1
+  if (output.sizes().size() != 0) {
+    output = at::slice(output, dim, 0, lengthScalar);
+  }
+  if (return_counts)
+    counts = at::slice(counts, 0, 0, lengthScalar);
+
+  return std::make_tuple(output, inverse_indices, counts);
+}
+
+std::tuple<Tensor, Tensor, Tensor>
+unique_consecutive_mps(const Tensor& self, const bool return_inverse, const bool return_counts, c10::optional<int64_t> dim) {
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("MPS: unique_consecutive op is supported natively starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performace implications.");
+    return at::unique_consecutive(self.to("cpu"), return_inverse, return_counts, dim);
+  }
+
+  return _unique_impl_mps(self, return_inverse, return_counts, true, dim);
+}
+
+std::tuple<Tensor, Tensor, Tensor>
+unique_dim_consecutive_mps(const Tensor& self, int64_t dim, const bool return_inverse, const bool return_counts) {
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("MPS: unique_dim_consecutive op is supported natively starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performace implications.");
+    return at::unique_dim_consecutive(self.to("cpu"), dim, return_inverse, return_counts);
```

Comment:
Same as above, result tensors must be copied back to MPS

Reply:
Updated